### PR TITLE
ATS

### PR DESCRIPTION
**Pull request by white238**
This PR handles adding ATS as a devtool and adds a basic ATS test workflow.

Important files:

`tests/integration/ats.sh.in`: Helper script that gets generated at the CMake step to run ats from the build directory
`tests/integration/test.ats`: Base ATS test file that exposes helpful variables and functions.  This also loads all subdirectory's `*.ats` files that it finds.
`tests/integration/default/default.ats`: adds a serial and parallel integration test for `data/input_files/default.lua`
`tests/integration/default/default_serial.json`: serial baseline summary file
`tests/integration/default/default_parallel.json`: parallel baseline summary file

Test repo PR: https://github.com/LLNL/serac_tests/pull/1

Relevant docs:
https://serac.readthedocs.io/en/task-white238-ats_devtool/sphinx/dev_guide/codevelop.html
https://serac.readthedocs.io/en/task-white238-ats_devtool/sphinx/dev_guide/testing.html

~~ATS Spack build needs this PR (just using the fork for now until this gets merged):
https://github.com/spack/spack/pull/27798~~
~~which also forces this to be fixed:
https://github.com/LLNL/uberenv/issues/73~~

- [x] Update TPL spack sha to past where the new uberenv needs
- [x] Build devtools on all platforms in shared location
- [x] Rebuild/hack regular host-configs to use new devtools
- [ ] Add rest of input-files
- [ ] Force @jamiebramwell to approve baselines
- [x] Figure out gitlab allocation scheme

### CODE REVIEW

File: scripts/spack/devtools.json [hunk 1]
```diff
@@ -3,9 +3,10 @@
 "package_version" : "fakeversion",
 "package_source_dir" : "../..",
 "package_final_phase" : "",
-"spack_url": "https://github.com/spack/spack",
-"spack_commit": "e75b76f433016a720027b41c0d36cdb9ef24a827",
+"spack_url": "https://github.com/adamjstewart/spack.git",
```
Comment (white238) [inline scripts/spack/devtools.json]: This will hopefully get merged very soon as we can go back to the normal repo for the devtools.

**Comment by white238** [comment]
@jamiebramwell  and @samuelpmishLLNL Can you give these two PRs a once over before I add the rest of the input files in `data/input_files`?

File: scripts/spack/configs/blueos_3_ppc64le_ib_p9/packages.yaml [hunk 2]
```diff
@@ -41,16 +41,6 @@ packages:
     variants: ~dbm
```
Comment (white238) [inline scripts/spack/configs/blueos_3_ppc64le_ib_p9/packages.yaml:41]: This is just updating this file to be inline with the devtools file and alphabetizing

File: scripts/spack/configs/mirrors.yaml [hunk 3]
```diff
@@ -0,0 +1 @@
+mirrors:: {}
```
Comment (white238) [inline scripts/spack/configs/mirrors.yaml:1]: I was running into a problem where spack's external http mirror wouldnt download to LC. This removes that.

File: scripts/spack/devtools_configs/blueos_3_ppc64le_ib_p9/packages.yaml [hunk 4]
```diff
@@ -20,123 +20,110 @@ packages:
       blas: [netlib-lapack]
```
Comment (white238) [inline scripts/spack/devtools_configs/blueos_3_ppc64le_ib_p9/packages.yaml:20]: This was very out of date and wasnt compliant with the clingo concretizer..

File: scripts/spack/devtools_configs/config.yaml [hunk 5]
```diff
@@ -0,0 +1 @@
+../configs/config.yaml
```
Comment (white238) [inline scripts/spack/devtools_configs/config.yaml:1]: symlink because apparently this got moved/was never there.. not sure

**Comment by jamiebramwell** [comment]
So if I'm reading this correctly, the baselines are stored in `serac/tests/integration/...` and the input files are stored in `serac/data/input_files/...`. Is it possible to have the test input files also in the test repo? This might be the long term plan and this is just a stopgap.

**Comment by white238** [comment]
> So if I'm reading this correctly, the baselines are stored in `serac/tests/integration/...` and the input files are stored in `serac/data/input_files/...`. Is it possible to have the test input files also in the test repo? This might be the long term plan and this is just a stopgap.

My idea was there was going to be two levels of integration test input files:

* ones meant as good examples for users, those go in `serac/data/input_files`
* ones not really meant for users for base consumption due to having to do "interesting" things to force test cases, those can go in `serac/tests/integration`

but both of these should be ran through ATS for testing.

**Comment by jamiebramwell** [comment]
> > So if I'm reading this correctly, the baselines are stored in `serac/tests/integration/...` and the input files are stored in `serac/data/input_files/...`. Is it possible to have the test input files also in the test repo? This might be the long term plan and this is just a stopgap.
> 
> My idea was there was going to be two levels of integration test input files:
> 
>     * ones meant as good examples for users, those go in `serac/data/input_files`
> 
>     * ones not really meant for users for base consumption due to having to do "interesting" things to force test cases, those can go in `serac/tests/integration`
> 
> 
> but both of these should be ran through ATS for testing.

Makes sense. Thanks for the explanation. Everything else looks good! I'll try installing ats locally and see how it goes.

**Comment by jamiebramwell** [comment]
Can we also install `ats` via the `+devtools` variant?

**Comment by white238** [comment]
> Can we also install `ats` via the `+devtools` variant?

Yes, with the normal spack caveat.  If you have an external package for them defined in your `packages.yaml`, spack will use that instead of installing them.  If you do not have the devtool listed it will install it with the variant.

**Comment by white238** [comment]
> > > So if I'm reading this correctly, the baselines are stored in `serac/tests/integration/...` and the input files are stored in `serac/data/input_files/...`. Is it possible to have the test input files also in the test repo? This might be the long term plan and this is just a stopgap.
> > 
> > 
> > My idea was there was going to be two levels of integration test input files:
> > ```
> > * ones meant as good examples for users, those go in `serac/data/input_files`
> > 
> > * ones not really meant for users for base consumption due to having to do "interesting" things to force test cases, those can go in `serac/tests/integration`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but both of these should be ran through ATS for testing.
> 
> Makes sense. Thanks for the explanation. Everything else looks good! I'll try installing ats locally and see how it goes.

We could maybe move them down the line to be in an `examples` directory imo.

**Comment by jamiebramwell** [comment]
This is now working on my local machine. Thanks @white238 !!!

File: scripts/spack/configs/linux_ubuntu_20/packages.yaml [hunk 6]
```diff
@@ -81,7 +81,11 @@ packages:
     externals:
     - spec: cuda@11.2.0
       prefix: /usr/local/cuda-11.2
-
+  m4:
```
Comment (jamiebramwell) [inline scripts/spack/configs/linux_ubuntu_20/packages.yaml:84]: Adding this to our ubuntu external libs sped things up considerably.

**Review by jamiebramwell — APPROVED** [review]

**Comment by white238** [comment]
https://github.com/LLNL/ATS/issues/68 to track ATS running parallel jobs on laptops.

File: host-configs/quartz-toss_3_x86_64_ib-clang@10.0.0.cmake [hunk 7]
```diff
@@ -97,7 +97,9 @@ set(SUNDIALS_DIR "${TPL_ROOT}/sundials-5.6.1" CACHE PATH "")
 
 # Root directory for generated developer tools
 
-set(DEVTOOLS_ROOT "/usr/WS2/smithdev/devtools/toss_3_x86_64_ib/2020_05_12_15_09_12/gcc-8.1.0" CACHE PATH "")
+set(DEVTOOLS_ROOT "/usr/WS2/smithdev/devtools/toss_3_x86_64_ib/2022_01_11_14_03_17/gcc-8.1.0" CACHE PATH "")
+
+set(ATS_EXECUTABLE "${DEVTOOLS_ROOT}/py-ats-7.0.10/bin/ats" CACHE PATH "")
```
Comment (jamiebramwell) [inline host-configs/quartz-toss_3_x86_64_ib-clang@10.0.0.cmake:102]: 👍 I think this is much less error prone.

**Review by jamiebramwell — APPROVED** [review]

**Review by samuelpmishLLNL — APPROVED** [review]

**Pull request merged**